### ISSUE TITLE
javascript is inserted in middle of page

### ISSUE THREAD

**Issue by wernerkrauss** [gordonbanderson/Mappable#8]
this slows down the page, as all js is loaded in middle of the page.

Is there a way to pass the generated JS to Requirements class to include it in bottom?

**Comment by gordonbanderson** [comment]
Previously all of the JavaScript was inline, I refactored out what I could into maputil.js.  The JavaScript that appears inline is what I saw as a minimum (namely the data to render and a call to maputil to create the map).
Options?
1) Generate JavaScript files on the fly into some temp space and include them at the bottom of the page via requirements
2) Defer the existing inline JavaScript to the bottom of the page.
3) Load POI data using an AJAX call, though this would probably slow things down further.

**Comment by wernerkrauss** [comment]
I think more of splitting up the template that has js and html in it and use http://api.silverstripe.org/3.1/class-Requirements_Backend.html#_javascriptTemplate in hope it adds the js to the bottom.

The call for generating the map is done after the page is rendered..?

**Comment by gordonbanderson** [comment]
In GoogleJavascript.ss

```
<script type="text/javascript">google.maps.event.addDomListener(window, 'load', loadedGoogleMapsAPI);</script>
```

**Comment by wernerkrauss** [comment]
The problem will remain until you use Requirements class to insert the javascript. if you have a script tag in a template all js is outputted there instead of the footer.

**Comment by gordonbanderson** [comment]
Changes pushed to https://github.com/gordonbanderson/Mappable/tree/ISSUE8

**Comment by gordonbanderson** [comment]
Changes pushed to 3.1 branch.  Had a few issues with short code street view and google maps but working now

**Comment by gordonbanderson** [comment]
Pushing the JavaScript to the foot of the HTML body breaks caching unfortunately.  Under the previous setup:

```
<% cached 'examplecachingofmarkersonamap', ID, LastEdited, Marker.LastEdited >
$Map
<% end_cached %>
```

Previously the JSON for the markers was inside the cached section but now it's not.

**Comment by wernerkrauss** [comment]
good point. The markers don't belong to the footer, and partial cache afaik just caches the result, not other calls like Requirement::javascript().

Maybe split it up to two calls? like 
```
<% cached (...) %>
$Map('false') //without javascript but markers
<% end_cached %>

$Map.RequireJavascript
```

Do the markers as JSOM also break javascript inclusion in footer?

**Comment by gordonbanderson** [comment]
I've pushed what I think is a fix for this to branch ISSUE8b - maps generated now render correctly when partially cached.  I had to move the register* methods into inline JavaScript in order to avoid loading the file maputil.js too early.
